pkg/lca/ibu_test.go

```diff
@@ -0,0 +1,693 @@
+package lca
+
+import (
+	"fmt"
+	"testing"
+
+	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+)
+
+var (
+	ibulGvk = schema.GroupVersionKind{
+		Group:   "lca.openshift.io",
+		Version: "v1alpha1",
+		Kind:    "ImageBasedUpgrade",
+	}
+)
+
+func TestImageBasedUpgradeWithOptions(t *testing.T) {
+	testSettings := buildTestClientWithDummyObject()
+	testBuilder, _ := PullImageBasedUpgrade(testSettings)
+	testBuilder = testBuilder.WithOptions(
+		func(builder *ImageBasedUpgradeBuilder) (*ImageBasedUpgradeBuilder, error) {
+			return builder, nil
+		})
+
+	assert.Equal(t, "", testBuilder.errorMsg)
+
+	testBuilder = testBuilder.WithOptions(
+		func(builder *ImageBasedUpgradeBuilder) (*ImageBasedUpgradeBuilder, error) {
+			return builder, fmt.Errorf("error")
+		})
+	assert.Equal(t, "error", testBuilder.errorMsg)
+}
+
+func TestImageBasedUpgradePull(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
+	}{
+		{
+			expectedError:       nil,
+			addToRuntimeObjects: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var (
+			runtimeObjects []runtime.Object
+			testSettings   *clients.Settings
+		)
+
+		testIBU := generateImageBasedUpgrade()
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testIBU)
+		}
+
+		testSettings = clients.GetTestClients(clients.TestClientParams{
+			K8sMockObjects: runtimeObjects,
+		})
+
+		// Test the PullImageBasedUpgrade function
+		builderResult, err := PullImageBasedUpgrade(testSettings)
+
+		// Check the error
+		assert.Equal(t, err, testCase.expectedError)
+
+		if testCase.expectedError == nil {
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeUpdate(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
+	}{
+		{
+			expectedError:       nil,
+			addToRuntimeObjects: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var (
+			runtimeObjects []runtime.Object
+			testSettings   *clients.Settings
+		)
+
+		testIBU := generateImageBasedUpgrade()
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testIBU)
+		}
+
+		testSettings = clients.GetTestClients(clients.TestClientParams{
+			K8sMockObjects: runtimeObjects,
+		})
+
+		ibuBuilder, err := PullImageBasedUpgrade(testSettings)
+		assert.Nil(t, err)
+
+		// Test the PullImageBasedUpgrade function
+		builderResult, err := ibuBuilder.WithSeedImage("quay.io/no-image").Update()
+
+		// Check the error
+		assert.Equal(t, err, testCase.expectedError)
+
+		if testCase.expectedError == nil {
+			assert.NotNil(t, builderResult)
+		}
```

Comment:
I think you should validate the value of spec.seedImageRef.image is what you set it to

Reply:
Added